third_party/move/move-compiler-v2/src/pipeline/reference_safety_processor.rs

```diff
@@ -543,667 +812,699 @@ enum ReadMode {
+                    );
+                    false
+                },
+                ReadMode::Argument => {
+                    // Mutable borrow not allowed
+                    if self.state.has_mut_edges(label) {
+                        self.error_with_hints(
+                            loc,
+                            format!(
+                                "cannot pass {} which is still mutably \
+                                    borrowed as function argument",
+                                self.display(local)
+                            ),
+                            "passed here",
+                            self.borrow_info(label, |_| true)
+                                .into_iter()
+                                .chain(usage_info),
+                        );
+                        false
+                    } else {
+                        true
+                    }
+                },
             }
+        } else {
+            true
         }
-        diags
     }
 
-    fn check_immut_edge(
-        &self,
-        state: &LifetimeState,
-        label: &LifetimeLabel,
-        edge: &BorrowEdge,
-    ) -> Vec<(Loc, String)> {
-        let mut diags = vec![];
-        // There must be no overlapping mutable child.
-        for (_, e) in state.effective_children(label) {
-            if e.is_mut && e.kind.overlaps(&edge.kind) {
-                if let Some(diag) = self.borrow_edge_info("previous ", state, true, e) {
-                    diags.push(diag)
+    /// Check whether a local can be written. This is only allowed if no borrowed references exist.
+    fn check_write_local(&self, local: TempIndex) {
+        if self.is_ref(local) {
+            // Always valid
+            return;
+        }
+        if let Some(label) = self.state.label_for_local_with_children(local) {
+            // The destination is currently borrowed and cannot be assigned
+            self.error_with_hints(
+                self.cur_loc(),
+                format!("cannot assign to borrowed {}", self.display(local)),
+                "attempted to assign here",
+                self.borrow_info(label, |_| true)
+                    .into_iter()
+                    .chain(self.usage_info(label, |t| t != &local)),
+            )
+        }
+    }
+
+    /// Check whether the borrow graph is 'safe' w.r.t a set of `locals`. See the discussion of safety at the
+    /// beginning of this file.
+    ///
+    /// To effectively check the path-oriented conditions of safety here, we need to deal with the fact
+    /// that graphs have non-explicit choice nodes, for example:
+    ///
+    /// ```ignore
+    ///                 s
+    ///            &mut /\ &mut
+    ///             .f /  \ .g
+    ///              r1    r2
+    /// ```
+    ///
+    /// The diverging edges `.f` and `.g` are not directly visible. In order to deal with this, we construct a
+    /// _hyper graph_ on the fly as follows:
```

Comment:
Is it necessary that this be a hypergraph?  Why not add an intermediate node corresponding to `&mut s`, with distinct out-edges labeled `.f` and `.g` to `r1` and `r2`, respectively?  Wouldn't that simplify your formulation, or would it complicate the expression of the safety constraints?